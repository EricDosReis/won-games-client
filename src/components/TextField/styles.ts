import styled, { css, DefaultTheme } from 'styled-components';

import { TextFieldProps } from '.';

type WrapperProps = {
  error: boolean;
} & Pick<TextFieldProps, 'iconPosition' | 'disabled'>;

const wrapperModifiers = {
  left: (theme: DefaultTheme) => css`
    ${Icon} {
      margin-right: ${theme.spacings.xsmall};
      order: 0;
    }
  `,

  right: (theme: DefaultTheme) => css`
    ${Icon} {
      margin-left: ${theme.spacings.xsmall};
      order: 1;
    }
  `,

  error: (theme: DefaultTheme) => css`
    ${Label},
    ${Icon} {
      color: ${theme.colors.red};
    }

    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }
  `,

  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, iconPosition, disabled, error }) => css`
    ${!!iconPosition && wrapperModifiers[iconPosition](theme)};
    ${disabled && wrapperModifiers.disabled(theme)};
    ${error && wrapperModifiers.error(theme)};
  `}
`;

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    background: ${theme.colors.lightGray};
    border-radius: ${theme.border.radius};
    border: 0.2rem solid ${theme.colors.lightGray};
    display: flex;
    padding: 0 ${theme.spacings.xsmall};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    background: transparent;
    border: 0;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    outline: none;
    padding: ${theme.spacings.xxsmall} 0;
    width: 100%;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    cursor: pointer;
    font-size: ${theme.font.sizes.small};
  `}
`;

export const Icon = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.gray};
    display: flex;
    align-items: center;
    justify-content: center;
    width: 2.2rem;

    svg {
      width: 2.2rem;
    }
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`;
