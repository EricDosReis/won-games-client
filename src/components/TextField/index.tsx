import { useState, InputHTMLAttributes } from 'react';

import * as S from './styles';

export type TextFieldProps = {
  onInput?: (value: string) => void;
  label?: string;
  initialValue?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
  disabled?: boolean;
  error?: string;
} & Omit<InputHTMLAttributes<HTMLInputElement>, 'onInput'>;

const TextField = ({
  onInput,
  label,
  name,
  initialValue = '',
  icon,
  iconPosition = 'left',
  disabled = false,
  error,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value;

    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  return (
    <S.Wrapper iconPosition={iconPosition} disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}

      <S.InputWrapper>
        {!!icon && <S.Icon>{icon}</S.Icon>}

        <S.Input
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>

      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
};

export default TextField;
