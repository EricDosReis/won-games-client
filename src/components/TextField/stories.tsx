import { Story, Meta } from '@storybook/react/types-6-0';
import { Email } from '@styled-icons/material-outlined';
import TextField, { TextFieldProps } from '.';

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' },
  },
} as Meta;

export const Default: Story<TextFieldProps> = args => (
  <div style={{ maxWidth: '30rem', padding: '1.5rem' }}>
    <TextField {...args} />
  </div>
);

export const Icon: Story<TextFieldProps> = args => (
  <div style={{ maxWidth: '30rem', padding: '1.5rem' }}>
    <TextField {...args} />
  </div>
);

Icon.args = {
  icon: <Email />,
};

export const Error: Story<TextFieldProps> = args => (
  <div style={{ maxWidth: '30rem', padding: '1.5rem' }}>
    <TextField {...args} error="Ops ... something is wrong" />
  </div>
);
