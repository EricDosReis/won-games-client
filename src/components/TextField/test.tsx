import { Email } from '@styled-icons/material-outlined';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import TextField from '.';

describe('<TextField />', () => {
  it('should render with label', () => {
    const { container } = renderWithTheme(
      <TextField label="Label" labelFor="Field" id="Field" />,
    );

    expect(screen.getByLabelText('Label')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
  });

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />);

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument();
  });

  it('it should change its value when typing', async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />,
    );

    const input = screen.getByRole('textbox');
    const text = 'This is my new text';

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });
    expect(onInput).toHaveBeenCalledWith(text);
  });

  it('should not change the value when disabled', async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
        disabled
      />,
    );

    const input = screen.getByRole('textbox');

    expect(input).toBeDisabled();

    const text = 'This is my new text';

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it('should not be accessible by tab when disabled', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        disabled
      />,
    );

    const input = screen.getByLabelText('TextField');

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).not.toHaveFocus();
  });

  it('should be accessible by tab', () => {
    renderWithTheme(
      <TextField label="TextField" labelFor="TextField" id="TextField" />,
    );

    const input = screen.getByLabelText('TextField');

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).toHaveFocus();
  });

  it('should render with icon', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
      />,
    );

    const icon = screen.getByTestId('icon');

    expect(icon).toBeInTheDocument();
    expect(icon.parentElement).toHaveStyle({
      marginRight: '1.6rem',
      order: '0',
    });
  });

  it('should render with icon on the right', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
        iconPosition="right"
      />,
    );

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      marginLeft: '1.6rem',
      order: '1',
    });
  });

  it('should render with error', () => {
    const { container } = renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        error="Error"
      />,
    );

    expect(screen.getByText(/error/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
