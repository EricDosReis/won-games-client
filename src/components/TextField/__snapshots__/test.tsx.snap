// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with error 1`] = `
.c0 .sc-eCssSg {
  margin-right: 1.6rem;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c0 .c1,
.c0 .sc-eCssSg {
  color: #FF6347;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c4 {
  background: #EAEAEA;
  border-radius: 0.4rem;
  border: 0.2rem solid #EAEAEA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1.6rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c5 {
  background: transparent;
  border: 0;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  outline: none;
  padding: 0.8rem 0;
  width: 100%;
}

.c2 {
  color: #030517;
  cursor: pointer;
  font-size: 1.4rem;
}

.c6 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="field"
  >
    Text Field
  </label>
  <div
    class="c3 c4"
  >
    <input
      class="c5"
      id="field"
      name="field"
      type="text"
      value=""
    />
  </div>
  <p
    class="c6"
  >
    Error
  </p>
</div>
`;

exports[`<TextField /> should render with label 1`] = `
.c0 .sc-eCssSg {
  margin-right: 1.6rem;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c2 {
  background: #EAEAEA;
  border-radius: 0.4rem;
  border: 0.2rem solid #EAEAEA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1.6rem;
}

.c2:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c3 {
  background: transparent;
  border: 0;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  outline: none;
  padding: 0.8rem 0;
  width: 100%;
}

.c1 {
  color: #030517;
  cursor: pointer;
  font-size: 1.4rem;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="field"
  >
    Label
  </label>
  <div
    class="c2"
  >
    <input
      class="c3"
      id="field"
      name="field"
      type="text"
      value=""
    />
  </div>
</div>
`;
